### PR TITLE
Fix zoom overlay positioning (Fixes #359)

### LINKED ISSUE #359: Zoom overlay drawn beneath shapes.
As demonstrated [here](https://plot.ly/~empet/13710/hexbin/), when drag-zooming, the shaded overlay is drawn beneath shapes.

### PR DESCRIPTION
**Pull request by n-riesco**
This PR fixes #359 by moving the select and zoombox elements from the subplot svg to the shapelayer.

@etpinard @mdtusz do you have suggestions for any tests I could add?

### CODE REVIEW

File: src/components/shapes/index.js [hunk 1]
```diff
@@ -115,86 +114,114 @@ shapes.add = function(gd) {
...
+    gd._fullLayout.shapes.splice(index, 1);
+
+    gd.layout.shapes.splice(index, 1);
+
+    for(var i = index; i < gd._fullLayout.shapes.length; i++) {
+        // redraw all shapes past the removed one,
+        // so they bind to the right events
+        gd._fullLayout._shapelayer
+            .selectAll('[data-index="' + (i+1) + '"]')
+            .attr('data-index', String(i));
```
Comment (mdtusz) [inline src/components/shapes/index.js]: FYI `String(i)` is unnecessary - `i` will be coerced to a string by `.attr`.

File: src/plots/cartesian/graph_interact.js [hunk 2]
```diff
@@ -1433,6 +1433,9 @@ function dragBox(gd, plotinfo, x, y, w, h, ns, ew) {
 
     var x0,
         y0,
+        xs = plotinfo.x()._offset,
```
Comment (mdtusz) [inline src/plots/cartesian/graph_interact.js]: 🐄  make the definitions before, and declarations after. 

e.g.

```
var xs = plotinfo.x()._offset,
    ys = plotinfo.y(),_offset,
    x0,
    y0,
    // etc.
```

**Comment by mdtusz** [comment]
Looks good so far. For testing, in `test/jasmine/assets/mouse_event.js` you'll find a method for "faking" mouse events. One test could be to simulate a click&drag zoom using that, and verify that the zoom overlay and shapes are drawn in the correct layers.

e.g.

``` js
// initiate the click&drag
mouseEvent('mousemove', startX, startY);
mouseEvent('mousedown', startX, startY);
mouseEvent('mousemove', endX, endY);

// check some things

// mouseup to finish the action
mouseEvent('mouseup', toX, toY);

// maybe test some more stuff
```

**Comment by n-riesco** [comment]
@mdtusz Thanks for the feedback!

I will add a test to `select_test.js` and create `zoom_test.js`.

**Comment by etpinard** [comment]
Moving the select and zoombox elements from the subplot svg to the shapelayer is only a short term solution before @mdtusz un-nests the subplot `<svg>` nodes. Looks like it does solves issue #359, so @n-riesco :beers: for that. 

That said, I don't think we should go forward with this PR - as it is only a temporary solution that may lead to confusion before we un-nests the subplot <svg> nodes. 

I can imagine @alexcjohnson saying:

> Wait, the zoomlayer is in the shape layer? Why??

Commits: https://github.com/plotly/plotly.js/pull/395/commits/df69584e893307e8e3ad351c3627fa2637357835 and https://github.com/plotly/plotly.js/pull/395/commits/e342a9e525ef00bec5e18e64ed832764a629b128 are valuable additions though.

**Comment by n-riesco** [comment]
@etpinard I've updated #378 with https://github.com/plotly/plotly.js/pull/395/commits/e342a9e525ef00bec5e18e64ed832764a629b128 , or do you think it's better I submit a new PR?

**Comment by alexcjohnson** [comment]
@n-riesco I like the way you've put the zoombox in a root layer and added a transform to it; but lets just make it a totally new root layer. I suspect it wants to go above [infolayer](https://github.com/plotly/plotly.js/blob/master/src/plot_api/plot_api.js#L2634) too, annotations likely have the same problem as shapes at the moment? I suppose we could put it in hoverlayer, since hover info is deleted as soon as zooming starts... but it'd be clearer and more flexible just to give it its own layer.

**Comment by n-riesco** [comment]
@alexcjohnson OK. I guess that once @mdtusz is done with the un-nesting, I can rebase this PR and create a new zoom layer.

**Comment by etpinard** [comment]
@n-riesco does https://github.com/plotly/plotly.js/pull/439 make this PR obsolete?

**Comment by n-riesco** [comment]
@etpinard No, it doesn't. I will rebase this PR, once #439 is merged. I don't expect the rebase to be difficult.

**Comment by n-riesco** [comment]
Moved to #448

**Pull request closed without merging**